Allow Docker image file permissions to be modified at runtime

JiamingMai opened the pull request:
### What changes are proposed in this pull request?

Rrefactor the Dockerfile's entrypoint script so that we can set permissions on a provided env/argument instead.

### Why are the changes needed?

Please clarify why the changes are needed. For instance,

Currently in the Alluxio Dockerfile the USER, UID and GID are all hard-coded. Thus only way to change the file permissions is to recompile a Docker image with new values for the arguments.

This creates a problem for users of the Helm chart who try to modify the user / UID used to run the Docker container in k8s. The corresponding file permissions do not get updated to the new UID.

I suggest we should consider refactoring the Dockerfile's entrypoint script so that we can set permissions on a provided env/argument instead.

### Does this PR introduce any user facing changes?

No, it doesn't.

beinan commented on integration/docker/Dockerfile:
```diff
@@ -25,9 +25,10 @@ RUN cd /alluxio-csi && \
+FROM alpine:3.10.8 AS alluxio-extractor
 # Note that downloads for *-SNAPSHOT tarballs are not available.
-ARG ALLUXIO_TARBALL=http://downloads.alluxio.io/downloads/files/2.9.0-SNAPSHOT/alluxio-2.9.0-SNAPSHOT-bin.tar.gz
+ARG ALLUXIO_TARBALL=https://downloads.alluxio.io/downloads/files/2.8.0/alluxio-2.8.0-bin.tar.gz
```
I think we need to keep the version to be 2.9.0-SNAPSHOT, which would be got updated for every release

beinan commented on integration/docker/Dockerfile-dev:
```diff
@@ -28,9 +28,9 @@ RUN cd /alluxio-csi && \
+FROM alpine:3.10.8 AS alluxio-extractor
 # Note that downloads for *-SNAPSHOT tarballs are not available.
-ARG ALLUXIO_TARBALL=http://downloads.alluxio.io/downloads/files/2.9.0-SNAPSHOT/alluxio-2.9.0-SNAPSHOT-bin.tar.gz
+ARG ALLUXIO_TARBALL=https://downloads.alluxio.io/downloads/files/2.8.0/alluxio-2.8.0-bin.tar.gz
```
keep 2.9.0-SNAPSHOT

beinan reviewed (changes requested):
looks good to me except the alluxio version

ssz1997 reviewed (commented):
Looks like the docker default user now is root. According to @ZhuTopher we can do something like `su <username>` in the entrypoint.sh?

JiamingMai commented:
I have changed the alluxio tarball version from 2.8.0 to 2.9.0-SNAPSHOT. The alpine's version is also changed back to 3.10.2. https://github.com/Alluxio/alluxio/pull/15898/commits/4441c096fc504a33cb6106b121fdae7a71970e0e

beinan reviewed (approved):
lgtm

ssz1997 commented:
@JiamingMai After I run `docker exec -it alluxio-master bash` and `whoami` inside the container, it returns user `root`. Is it possible to run as user Alluxio by default  or the user specified through the environment variable? 
```
bash-5.0# ps aux
PID   USER     TIME  COMMAND
    1 a         0:00 bash /entrypoint.sh master
   74 root      0:00 bash
  326 a         0:12 /usr/lib/jvm/java-1.8-openjdk/bin/java -cp /opt/alluxio-2.9.0-SNAPSHOT/conf/::/opt/alluxio-2.9.0-SNAPSHOT/assembly/alluxio-server-2.9.0
  327 a         0:17 /usr/lib/jvm/java-1.8-openjdk/bin/java -cp /opt/alluxio-2.9.0-SNAPSHOT/conf/::/opt/alluxio-2.9.0-SNAPSHOT/assembly/alluxio-server-2.9.0
  667 root      0:00 ps aux
bash-5.0# whoami
root
```

beinan commented:
> @JiamingMai After I run `docker exec -it alluxio-master bash` and `whoami` inside the container, it returns user `root`. Is it possible to run as user Alluxio by default or the user specified through the environment variable?
> 
> ```
> bash-5.0# ps aux
> PID   USER     TIME  COMMAND
>     1 a         0:00 bash /entrypoint.sh master
>    74 root      0:00 bash
>   326 a         0:12 /usr/lib/jvm/java-1.8-openjdk/bin/java -cp /opt/alluxio-2.9.0-SNAPSHOT/conf/::/opt/alluxio-2.9.0-SNAPSHOT/assembly/alluxio-server-2.9.0
>   327 a         0:17 /usr/lib/jvm/java-1.8-openjdk/bin/java -cp /opt/alluxio-2.9.0-SNAPSHOT/conf/::/opt/alluxio-2.9.0-SNAPSHOT/assembly/alluxio-server-2.9.0
>   667 root      0:00 ps aux
> bash-5.0# whoami
> root
> ```

@JiamingMai , could you take a look at @ssz1997 's comments?  thanks!

JiamingMai commented:
> @JiamingMai After I run `docker exec -it alluxio-master bash` and `whoami` inside the container, it returns user `root`. Is it possible to run as user Alluxio by default or the user specified through the environment variable?
> 
> ```
> bash-5.0# ps aux
> PID   USER     TIME  COMMAND
>     1 a         0:00 bash /entrypoint.sh master
>    74 root      0:00 bash
>   326 a         0:12 /usr/lib/jvm/java-1.8-openjdk/bin/java -cp /opt/alluxio-2.9.0-SNAPSHOT/conf/::/opt/alluxio-2.9.0-SNAPSHOT/assembly/alluxio-server-2.9.0
>   327 a         0:17 /usr/lib/jvm/java-1.8-openjdk/bin/java -cp /opt/alluxio-2.9.0-SNAPSHOT/conf/::/opt/alluxio-2.9.0-SNAPSHOT/assembly/alluxio-server-2.9.0
>   667 root      0:00 ps aux
> bash-5.0# whoami
> root
> ```

You can docker exec -it -u user_name

ssz1997 reviewed (approved):
LGTM. Thanks for adding this functionality

beinan commented:
alluxio-bot, merge this please.

ZhuTopher commented:
Duplicating this comment for posterity: https://github.com/Alluxio/alluxio/pull/15980#discussion_r933496712

---

By the way forcing the Pod/Container to use the root user from https://github.com/Alluxio/alluxio/pull/15898 is slightly concerning from a security point-of-view (to put it lightly) 😅

Can we refactor the Dockerfile to create a different admin user with permission to create users/groups & assign file ownership (i.e: a sudo-er), but isn't root (uid=0, gid=0)? The rest will still be the same; after creating the alluxio user & group you run the process as that user. This is just so that the container image itself doesn't run with a root user (which is very dangerous if people can enter the container runtime somehow).

Example blogpost: https://blog.aquasec.com/rootless-containers-boosting-container-security
